From f7d1f552f6a148d90dbdd646aecf8224501f4523 Mon Sep 17 00:00:00 2001
Message-ID: <f7d1f552f6a148d90dbdd646aecf8224501f4523.1782305243.git.pmatilai@redhat.com>
From: Panu Matilainen <pmatilai@redhat.com>
Date: Wed, 24 Jun 2026 15:39:24 +0300
Subject: [PATCH] Fix rubygem extract regression from
 b8b54f92c65bca7bfdd80c3ecf717d5c48b02e91

As b8b54f92c65bca7bfdd80c3ecf717d5c48b02e91 notes, this could've used
some C++ love and maybe avoided the regression too by being waaaayy simpler.
Adjust test expectation to pass now.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2492128
---
 tests/rpmmacro.at      |  1 -
 tools/rpmuncompress.cc | 17 ++++++-----------
 2 files changed, 6 insertions(+), 12 deletions(-)

diff --git a/tools/rpmuncompress.cc b/tools/rpmuncompress.cc
index dc49921bf..76ea6ab32 100644
--- a/tools/rpmuncompress.cc
+++ b/tools/rpmuncompress.cc
@@ -6,6 +6,7 @@
 #include <stdio.h>
 #include <string.h>
 
+#include <filesystem>
 #include <string>
 
 #include <archive.h>
@@ -17,6 +18,8 @@
 
 #include "debug.h"
 
+namespace fs = std::filesystem;
+
 static int extract = 0;
 static int dryrun = 0;
 static char *dstpath = NULL;
@@ -217,21 +220,13 @@ static char *doUntar(const char *fn)
 	if (needtar) {
 	    rasprintf(&buf, "%s %s '%s' | %s %s - %s", mkdir, zipper, fn, tar, taropts, stripcd);
 	} else if (at->compressed == COMPRESSED_GEM) {
-	    char *tmp = xstrdup(fn);
-	    const char *bn = basename(tmp);
-	    size_t nvlen = strlen(bn) - 3;
+	    auto bn = fs::path(fn).stem();
 	    char *gem = rpmGetPath("%{__gem}", NULL);
-	    char *gemspec = NULL;
-	    std::string gemnameversion(bn, nvlen);
-
-	    gemspec = rpmGetPath("", gemnameversion.c_str(), ".gemspec", NULL);
 
-	    rasprintf(&buf, "%s '%s' && %s spec '%s' --ruby > '%s'",
-			zipper, fn, gem, fn, gemspec);
+	    rasprintf(&buf, "%s '%s' && %s spec '%s' --ruby > '%s.gemspec'",
+			zipper, fn, gem, fn, bn.c_str());
 
-	    free(gemspec);
 	    free(gem);
-	    free(tmp);
 	} else {
 	    rasprintf(&buf, "%s%s '%s' %s", mkdir, zipper, fn, stripcd);
 	}
-- 
2.54.0

